### PR TITLE
Enhancement/tutor dashboard/support late submissions

### LINKED ISSUE #1376: Properly support late submissions in tutor exercise dashboard
The tutor exercise dashboard (e.g. https://artemis.ase.in.tum.de/#/course-management/48/exercises/1182/tutor-dashboard) does not really distinguish between submissions made before the due date (in time) and after the due date (late).

In time submissions lead to a rated result and should have a higher priority while grading, late submissions lead to an unrated result and should have a lower priorities while grading. So only if there are no in time submissions left to assess, tutors should be offered late submissions.

The tutor exercise dashboard should distinguish between in time and late submissions when displaying them to the tutors, e.g. by using a label (similar to a category in the exercise overview) that immediately shows if a submission was in time or late.

This should also be made clear in the statistics, so that no wrong statistics (negative counts, percentage values outside of the range [0%, 100%] are displayed. I think adding multiple rows for late submissions counts and in time submissions counts (or even a table) would make sense here to distinguish.

image

In the tutor course dashboard, exercises should only turn green if 100% of the in time submissions are assessed. For late submissions, it would make sense to add a separate progress bar. Again, it should be emphasized that the higher priority are in time submissions here!

image

Small additional request: for both, tutor exercise dashboard and tutor course dashboard, it would be nice to show a loading icons here until all data is loaded from the server.

### PR DESCRIPTION
**Pull request by schultek**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://vmbruegge60.in.tum.de
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Resolves #1376

The tutor dashboard (and other pages) do not distinguish between in-time submissions and late submissions for exercises. 
While this can lead to confusion or a lack of clarity for tutors, it also introduces statistical errors on certain calculations (see issue #1376)

### Description

1. Introduced separate variables for in-time and late submissions, as well as assessment counts of in-time and late submissions. In both java and js the class *DueDateStat* is used to give some structure. It contains an **in-time** value and a **late** value. In js, it also contains a **total** computed value.  
   Changes are made mainly to: 
  - [Java Models]: +DueDateStat.java, Exercise.java, StatsForInstructorDashboardDTO.java
  - [Java Api]: CourseResource.java, ExerciseResource.java
  - [JS]: +due-date-stat.model.ts, instructor-course-dashboard, tutor-course-dashboard, tutor-exercise-dashboard

 2. Extended the statistics table to properly show and calculate late submission stats in tutor-exercise-dashboard and instructor-exercise-dashboard.

 3. Extended the *tutor-participation-graph* by a separate progress bar for late submissions / assessments. This is shown parallel to the in-time bar and only, if there are any late submissions. The overall progress status (coloring) is not affected by the late submission status.

 4. Modify the assessment order, so that late-submissions are assessed last and only if there are no in-time submissions left to assess.
    - The Modeling Exercises have a Compass Service, which, when requesting a new assessment, calculates the submission with the most knowledge gain for automatic assessments. Currently this is done **before** doing the prioritization. Therefore it can sill happen that late submissions are selected for new assessments while there are unassessed in-time submissions left. 

### Steps for Testing

1. Log in as different students
2. Create in-time as well as late submissions for an exercise
3. Log in as a tutor
4. Choose a course with late submissions
5. Navigate to Course Administration -> Tutor Dashboard ( -> Exercise Dashboard )
6. Review: participation graph and statistics table
7. Check correct choosing of submissions for new assessments
8. Navigate to Course Administration -> Instructor Dashboard
9. Review: statistics table and exercises list

### Screenshots

Participation Graph
Bildschirmfoto 2020-05-18 um 16 54 37

### CODE REVIEW

**Comment by madwau** [comment]
@schultek Could you resolve the conflicts?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package de.tum.in.www1.artemis.web.rest.dto;
+
+public class DueDateStat {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java:3]: DueDateStatistic
--> JavaDoc Kommentare hinzufügen mit einer Erklärung wie das verwendet wird

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    private Long inTime;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java:5]: numberOfInTimeElement ?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    private Long late;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java:7]: numberOfLateElements?

**Comment by FuchsDominik** [comment]
Capture
Is this wanted behaviour?
This was the only odd thing that i experienced, the rest worked perfectly fine. I like that it does not show the branch when no submission are late.

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 4]
```diff
@@ -814,19 +815,19 @@ public void setTutorParticipations(Set<TutorParticipation> tutorParticipations)
...
-    public Long getNumberOfParticipations() {
-        return numberOfParticipationsTransient;
+    public DueDateStat getNumberOfSubmissions() {
+        return numberOfSubmissionsTransient;
     }
 
-    public void setNumberOfParticipations(Long numberOfParticipations) {
-        this.numberOfParticipationsTransient = numberOfParticipations;
+    public void setNumberOfSubmissions(DueDateStat numberOfnumberOfSubmissions) {
+        this.numberOfSubmissionsTransient = numberOfnumberOfSubmissions;
```
Comment (anditurdiu) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:823]: This is probably a typo? numberOfnumberOfSubmissions

**Comment by anditurdiu** [comment]
After testing on the test server 1, everything worked as expected for me! Good job!

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 4 above)
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java]: Done

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java (hunk 1 above)
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/DueDateStat.java:7]: Done

**Comment by schultek** [comment]
I cannot reproduce this. Does the exercise have any late submissions? If not, this is the wanted behavior.

**Comment by Varnika-Tyagi** [comment]
> Yes the exercise has 1 late submission. I added some screenshots for better clarification
> 
> ## Bildschirmfoto 2020-06-01 um 22 16 21
> ## Bildschirmfoto 2020-06-01 um 22 16 47
> Bildschirmfoto 2020-06-01 um 22 17 11

Facing the same issue as Mario. I don't see the 1 late submission.
TestingPR1485

**Comment by Varnika-Tyagi** [comment]
Works now! Well done!
RetestingPR1485
RetestingPR1485-1

**Pull request merged**